### PR TITLE
[CMS] Adds edit invoice feature

### PR DESCRIPTION
**Pull request by thi4go**
This PR still needs some testing, but implements edit invoice feature.

#1060

### CODE REVIEW

**Comment by alexlyp** [comment]
Can you rebase this please?

**Comment by thi4go** [comment]
Hey mates @alexlyp @fernandoabolafio this PR is ready to be reviewed.  Can you guys give it a go? I also had to check something with fernando but riot is not colaborating today 😅

**Comment by alexlyp** [comment]
Sure thing.  Can you take a look at #1084 ?

**Comment by alexlyp** [comment]
So looks like it updated properly for editing Contractor fields and any rows that were previously there, but adding new rows wasn't working and it kept saying it was a duplicate invoice (if i didn't change anything else).

**Comment by alexlyp** [comment]
I also noticed that the edit button needs to be only shown in these situations:
- Only show for invoices in New, Updated or Disputed statuses.  Anything else it shouldn't be able to edit.
- Only show for invoices owned by the current user (this may be tougher to do).

**Comment by alexlyp** [comment]
Also on Edit I noticed that a previously blacked out cell for expenses on a Labor (type 1) line item became editable.

**Comment by alexlyp** [comment]
Oh and a "cancel" button next to Update would be great as well.

**Comment by thi4go** [comment]
Right, working on the changes! Thanks for the review alex.

So is the `Expense` cell supposed to be Readonly at all times? I tested here and I'm actually getting that when creating new invoices as well. When I add a row, the Readonly cell shifts to the left it seems.

File: src/components/snew/RegisterForm.js [hunk 1]
```diff
@@ -22,7 +22,7 @@ const RegisterForm = ({
   handleSubmit,
   isCMS
 }) =>
-  loggedInAsEmail ? null : isCMS ? (
```
Comment (fernandoabolafio) [inline src/components/snew/RegisterForm.js]: I don't think the diffs in this file belongs to this PR, right? Maybe a rebasing issue.

File: src/validators/signup.js [hunk 2]
```diff
@@ -13,10 +13,7 @@ const validate = (policy, values, isCMS) => {
       !isRequiredValidator(values.verificationtoken) ||
       !isRequiredValidator(values.username) ||
       !isRequiredValidator(values.password) ||
-      !isRequiredValidator(values.password_verify) ||
```
Comment (fernandoabolafio) [inline src/validators/signup.js]: Looks to me that this diff doesn't belong to this PR. Maybe a rebasing issue?

File: src/selectors/app.js [hunk 3]
```diff
@@ -63,9 +62,11 @@ import {
 } from "../constants";
 import {
   getTextFromIndexMd,
+  getTextFromJsonToCsv,
   countPublicProposals,
   isProposalApproved
 } from "../helpers";
+import { convertLineItemsToGrid } from "../components/InvoiceDatasheet/helpers";
```
Comment (fernandoabolafio) [inline src/selectors/app.js]: I think you'll need to make this function more robust in order to update the datasheet cells accordingly. There are some rules which you can see when you're creating new invoices, such as:
- if type cell === 1 -> expense cell is read-only
- if type cell === 2 or 3 -> labor cell is read-only 

If this function is called with readOnly enable, this check is not necessary. Otherwise, we need to make sure this happens when populating the table.

There is already a function which updates a grid based on the value of a given cell, maybe we can integrate that into this function. See: https://github.com/decred/politeiagui/blob/master/src/components/InvoiceDatasheet/helpers.js#L294.

Comment (thi4go) [reply src/selectors/app.js]: Nice Fernando, didn't know about those restrictions, thanks for clarifying. Right now I pushed an easy fix to this problem, without using the `processCellsChange` function, though I did try to place it somewhere within that function and wasnt quite able to...

File: src/components/snew/RegisterForm.js (hunk 1 above)
Comment (thi4go) [reply src/components/snew/RegisterForm.js]: Trying to remove this from the PR, not sure how to exactly 🤔

File: src/validators/signup.js (hunk 2 above)
Comment (thi4go) [reply src/validators/signup.js]: Trying to remove this from the PR, not sure how to exactly 🤔

**Comment by alexlyp** [comment]
Everything looks good to me after testing.  Will let @fernandoabolafio do the final approval for any issues codewise. Great job!

**Comment by thi4go** [comment]
Hey @fernandoabolafio, thanks for the thorough review mate. So I investigated it a bit further, and I stumbled into some weird behaviour. So there is two blocks on this form, the first one with input fields (name, contact, rate...) and the grid with line items. I noticed that:

- if I edit information on the input fields AND the line items, only the line items gets updated correctly;
- the input fields gets updated correctly only if you don't change line items on the same request;

Do you guys have any suspicion on why this is happening? @fernandoabolafio @alexlyp I'll keep investigating further anyhow.

**Comment by fernandoabolafio** [comment]
Hey, @thi4go here is the video in another hosting service: https://sendvid.com/pck7wqaa.

About this I think is probably due to misuse of the connectors, but I haven't noticed that while testing the feature. The form updating looks fine to me as I type the new values. What I noticed as errored is what I captured in the video: the updated values in the form are not being sent to the API.
> Do you guys have any suspicion on why this is happening? @fernandoabolafio @alexlyp I'll keep investigating further anyhow.

**Comment by thi4go** [comment]
Hey @fernandoabolafio, the form is fixed now, thanks for the hint, it was indeed a problem with the connector. 👍
